### ISSUE TITLE
MQTT client fails to reconnect due to socket not available for writing

### ISSUE THREAD

**Issue by nirmitdesai** [eclipse/paho.mqtt.c#1232]
Hello,

We've few clients which connect to MQTT server using Paho MQTT c library. The clients are configured to automatically reconnect.
When the MQTT server is restarted, we see that connection is lost as expected. Once MQTT server is back up, expected behavior is that the automatic reconnect logic will kick in and ensure that the clients get reconnected. However, we are seeing that even after 3-4 hours, the clients haven't reconnected to MQTT server.

The original log file is 225 MB so I've zipped and uploaded it. If there's other alternative, please let me know.

Below is my analysis from the log file:

**First connection attempt was made:**

> 125:20220526 153330.255 Connecting to serverURI v2ls-edge-02.steslabs.net:31883 with MQTT version 5

**After a while, the connection was lost when MQTT server went down:**

> 2552094:20220528 002702.486 Calling connectionLost for client EDGE02VD2-1FS

**MQTT library automatically retried connection at exponentially increasing times as shown below:**

>2557083:20220528 002754.735 Connecting to serverURI v2ls-edge-02.steslabs.net:31883 with MQTT version 5
>2567766:20220528 003007.820 Connecting to serverURI v2ls-edge-02.steslabs.net:31883 with MQTT version 5
>2592645:20220528 003449.414 Connecting to serverURI v2ls-edge-02.steslabs.net:31883 with MQTT version 5
>2635220:20220528 004235.610 Connecting to serverURI v2ls-edge-02.steslabs.net:31883 with MQTT version 5
>2740387:20220528 010115.993 Connecting to serverURI v2ls-edge-02.steslabs.net:31883 with MQTT version 5
>2933840:20220528 013522.515 Connecting to serverURI v2ls-edge-02.steslabs.net:31883 with MQTT version 5
>3284691:20220528 023649.668 Connecting to serverURI v2ls-edge-02.steslabs.net:31883 with MQTT version 5
>3618768:20220528 033520.695 Connecting to serverURI v2ls-edge-02.steslabs.net:31883 with MQTT version 5
>4024143:20220528 044616.353 Connecting to serverURI v2ls-edge-02.steslabs.net:31883 with MQTT version 5

**We can see reconnection attempt made first at 00:27:54, then, after ~2 mins (00:30::07), then after ~4 mins (00:34:49), then after 8,16, 32 mins interval. Finally, reconnection attempt is made every hour (02:36:49, 03:35:20, 04:46:16).*8

However, in none of these 9 retry attempts, reconnection was successful.

**For every reconnection attempt, a new socket is used :**

>2557090:20220528 002754.735 New socket 2376 for v2ls-edge-02.steslabs.net:31883, port 31883
>2567773:20220528 003007.898 New socket 1272 for v2ls-edge-02.steslabs.net:31883, port 31883
>2592652:20220528 003449.414 New socket 1272 for v2ls-edge-02.steslabs.net:31883, port 31883
>2635227:20220528 004235.610 New socket 1168 for v2ls-edge-02.steslabs.net:31883, port 31883>
>2740394:20220528 010115.993 New socket 1340 for v2ls-edge-02.steslabs.net:31883, port 31883
>2933847:20220528 013522.515 New socket 1316 for v2ls-edge-02.steslabs.net:31883, port 31883
>3284698:20220528 023649.668 New socket 1740 for v2ls-edge-02.steslabs.net:31883, port 31883
>3618775:20220528 033520.758 New socket 1312 for v2ls-edge-02.steslabs.net:31883, port 31883
>4024150:20220528 044616.353 New socket 1380 for v2ls-edge-02.steslabs.net:31883, port 31883

MQTT library uses select API to check if the socket is writeable. The connect timeout is 30 seconds so MQTT keeps checking if socket is available for writing.

grep -in -A 800 "new socket 1168" mqtt_log.txt | grep -i "from write select"

```
2635270-20220528 004235.719 Return code 0 from write select
2635290-20220528 004236.828 Return code 0 from write select
2635333-20220528 004238.641 Return code 0 from write select
2635351-20220528 004239.063 Return code 0 from write select
2635369-20220528 004240.656 Return code 0 from write select
2635412-20220528 004241.657 Return code 0 from write select
2635432-20220528 004242.657 Return code 0 from write select
2635450-20220528 004244.516 Return code 0 from write select
2635493-20220528 004244.688 Return code 0 from write select
2635515-20220528 004246.734 Return code 0 from write select
2635558-20220528 004247.719 Return code 0 from write select
2635578-20220528 004247.844 Return code 0 from write select
2635596-20220528 004248.953 Return code 0 from write select
2635639-20220528 004250.734 Return code 0 from write select
2635657-20220528 004251.297 Return code 0 from write select
2635675-20220528 004252.422 Return code 0 from write select
2635720-20220528 004253.781 Return code 0 from write select
2635738-20220528 004254.781 Return code 0 from write select
2635756-20220528 004255.796 Return code 0 from write select
2635803-20220528 004257.812 Return code 0 from write select
2635821-20220528 004259.000 Return code 0 from write select
2635866-20220528 004300.109 Return code 0 from write select
2635884-20220528 004301.218 Return code 0 from write select
2635902-20220528 004301.218 Return code 0 from write select
2635945-20220528 004302.890 Return code 0 from write select
2635963-20220528 004303.562 Return code 0 from write select
2635983-20220528 004304.671 Return code 0 from write select
```

The same is true for all reconnection attempts.

In same environment, I had another client which was able to reconnect and main difference I could see was that socket was available for writing. So the SSL connection could be established again and client was reconnected. I've added a short snippet of those logs as well( successful_reconnect.txt).

Do you think increasing the connect timeout would help? Is there anything else that I might be missing?
Please let me know if you require any more information.

Thanks!

[mqtt_logs.zip](https://github.com/eclipse/paho.mqtt.c/files/8821632/mqtt_logs.zip)
[successful_reconnect.txt](https://github.com/eclipse/paho.mqtt.c/files/8821633/successful_reconnect.txt)

**Comment by icraggs** [comment]
The trace looks a bit strange when it's trying to reconnect. You could try using the command line utility paho_c_sub or paho_c_pub, just to see if the automatic reconnect works for them as it's supposed to. If it does, then there's a difference between your program and those. If it doesn't it might be easier to debug in a simpler program.

(You don't need to set the trace level to maximum for this purpose - the normal level will do and will keep the trace shorter).

**Comment by icraggs** [comment]
Probably a duplicate of #1215

**Comment by nirmitdesai** [comment]
Hi Ian, can you elaborate more about what part about trace looked strange? 
Since the issue happens at scale, I would need to see how to manage a big deployment of paho_c_sub/paho_c_pub .
What do you infer from the fact that the sockets are not available for writing on every reconnect attempt that is being made?

**Comment by icraggs** [comment]
What do you mean at scale? Number of clients, messages or what?

The trace is only for one client, is that right? You seem to have a reconnect attempt every 3 seconds, if I'm reading that right. That is strange. There's no chance for one reconnect attempt to get any where before the next one starts. Have you set the min and max reconnect intervals?

**Comment by icraggs** [comment]
If you take another trace for this, do it at the minimum level,  not maximum. Max level adds stuff we don't need in this case.

**Comment by nirmitdesai** [comment]
Sorry for the late reply.

> What do you mean at scale? Number of clients, messages or what?

By scale I meant number of clients. We see the issue happening when there are more than 15,000 clients.

> The trace is only for one client, is that right?

The trace is for only one client. 

> You seem to have a reconnect attempt every 3 seconds, if I'm reading that right. That is strange. There's no chance for one > reconnect attempt to get any where before the next one starts

Aren't below logs indicating reconnection attempt happening in exponential period:

> 2557083:20220528 002754.735 Connecting to serverURI v2ls-edge-02.steslabs.net:31883 with MQTT version 5
> 2567766:20220528 003007.820 Connecting to serverURI v2ls-edge-02.steslabs.net:31883 with MQTT version 5
> 2592645:20220528 003449.414 Connecting to serverURI v2ls-edge-02.steslabs.net:31883 with MQTT version 5
> 2635220:20220528 004235.610 Connecting to serverURI v2ls-edge-02.steslabs.net:31883 with MQTT version 5
> 2740387:20220528 010115.993 Connecting to serverURI v2ls-edge-02.steslabs.net:31883 with MQTT version 5
> 2933840:20220528 013522.515 Connecting to serverURI v2ls-edge-02.steslabs.net:31883 with MQTT version 5
> 3284691:20220528 023649.668 Connecting to serverURI v2ls-edge-02.steslabs.net:31883 with MQTT version 5
> 3618768:20220528 033520.695 Connecting to serverURI v2ls-edge-02.steslabs.net:31883 with MQTT version 5
> 4024143:20220528 044616.353 Connecting to serverURI v2ls-edge-02.steslabs.net:31883 with MQTT version 5

I am not sure if I interpreted it incorrectly.

> Have you set the min and max reconnect intervals?

Yes, min interval is set to 60 seconds and max interval to 3600 seconds.

> If you take another trace for this, do it at the minimum level, not maximum. Max level adds stuff we don't need in this case.

Sure, will do

**Comment by icraggs** [comment]
I've added a fix for #1215 which I think will fix this one too.

**Comment by nirmitdesai** [comment]
Thanks! I can test it out and update you.
